File: 23.Message-Brokers-and-Event-Streaming-with-Kafka/code/typescript/example_2.ts
import { Kafka } from 'kafkajs';

const kafka = new Kafka({
  clientId: 'my-app',
  brokers: ['localhost:9092']
});

// the consumer GROUP ,  scale by adding instances
const consumer = kafka.consumer({ groupId: 'billing' });

async function run() {
  await consumer.connect();
  
  // where to start if no committed offset (sec 8)
  await consumer.subscribe({ topic: 'orders', fromBeginning: true });

  await consumer.run({
    autoCommit: false, // we'll commit manually for at-least-once (sec 9)
    eachMessage: async ({ topic, partition, message }) => {
      // do the work FIRST...
      console.log(`Processing: ${message.value?.toString()}`);
      
      // ...THEN commit the offset (at-least-once, sec 9)
      // Note: kafkajs auto-commits if you don't throw an error in eachMessage, 
      // but to do it manually:
      /*
      await consumer.commitOffsets([{
        topic,
        partition,
        offset: (BigInt(message.offset) + 1n).toString()
      }]);
      */
    },
  });
}

run().catch(console.error);

File: 23.Message-Brokers-and-Event-Streaming-with-Kafka/code/rust/example_2.rs
use rdkafka::config::ClientConfig;
use rdkafka::consumer::{Consumer, StreamConsumer, CommitMode};
use rdkafka::Message;

#[tokio::main]
async fn main() {
    let consumer: StreamConsumer = ClientConfig::new()
        .set("bootstrap.servers", "localhost:9092")
        .set("group.id", "billing") // the consumer GROUP , scale by adding instances
        .set("auto.offset.reset", "earliest") // where to start if no committed offset (sec 8)
        .set("enable.auto.commit", "false") // we'll commit manually for at-least-once (sec 9)
        .create()
        .expect("Consumer creation failed");

    // Kafka assigns this instance some partitions
    consumer.subscribe(&["orders"]).expect("Can't subscribe");

    loop {
        // blocks for the next record
        match consumer.recv().await {
            Err(e) => println!("Kafka error: {}", e),
            Ok(m) => {
                let payload = match m.payload_view::<str>() {
                    None => "",
                    Some(Ok(s)) => s,
                    Some(Err(e)) => {
                        println!("Error while deserializing message payload: {:?}", e);
                        ""
                    }
                };
                
                // process(msg.Value) -> do the work FIRST...
                println!("Processing payload: {}", payload);

                // ...THEN commit the offset (at-least-once, sec 9)
                consumer.commit_message(&m, CommitMode::Async).unwrap();
            }
        };
    }
}
